File: api/upload-share-image.php
<?php
header('Content-Type: application/json');
header('Access-Control-Allow-Origin: *');
header('Access-Control-Allow-Methods: POST');
header('Access-Control-Allow-Headers: Content-Type');

if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    http_response_code(405);
    echo json_encode(['error' => 'Method not allowed']);
    exit;
}

if (!isset($_FILES['image']) || $_FILES['image']['error'] !== UPLOAD_ERR_OK) {
    http_response_code(400);
    echo json_encode(['error' => 'No image file or upload error']);
    exit;
}

$file = $_FILES['image'];
$uploadDir = __DIR__ . '/../share-images/';
$allowedTypes = ['image/png', 'image/jpeg', 'image/jpg'];

// Создаем директорию, если не существует
if (!is_dir($uploadDir)) {
    mkdir($uploadDir, 0755, true);
}

// Генерируем уникальное имя файла на основе параметров
$age = isset($_POST['age']) ? intval($_POST['age']) : 0;
$bioAge = isset($_POST['bioAge']) ? floatval($_POST['bioAge']) : 0;
$hash = md5($age . '_' . $bioAge . '_' . time());
$fileName = 'share_' . $hash . '.png';
$filePath = $uploadDir . $fileName;

// Проверяем тип файла
$finfo = finfo_open(FILEINFO_MIME_TYPE);
$mimeType = finfo_file($finfo, $file['tmp_name']);
finfo_close($finfo);

if (!in_array($mimeType, $allowedTypes)) {
    http_response_code(400);
    echo json_encode(['error' => 'Invalid file type: ' . $mimeType]);
    exit;
}

// Проверяем размер файла (макс 5MB)
if ($file['size'] > 5 * 1024 * 1024) {
    http_response_code(400);
    echo json_encode(['error' => 'File too large']);
    exit;
}

// Перемещаем файл
if (move_uploaded_file($file['tmp_name'], $filePath)) {
    // Устанавливаем правильные права
    chmod($filePath, 0644);
    
    // Определяем базовый URL
    $protocol = isset($_SERVER['HTTPS']) && $_SERVER['HTTPS'] === 'on' ? 'https' : 'http';
    $host = $_SERVER['HTTP_HOST'];
    $baseUrl = $protocol . '://' . $host;
    
    $publicUrl = $baseUrl . '/share-images/' . $fileName;
    echo json_encode(['url' => $publicUrl]);
} else {
    http_response_code(500);
    echo json_encode(['error' => 'Failed to upload file']);
}
?>

File: api/upload-share-image.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Серверный скрипт для загрузки изображений для шаринга
Использование: запустить через fcgiwrap (CGI)
"""

import os
import sys
import json
import cgi
import cgitb
import hashlib
from datetime import datetime
from pathlib import Path

# Включаем обработку ошибок для CGI
cgitb.enable()

# Настройки
UPLOAD_DIR = Path('/var/www/html/app/share-images')
MAX_FILE_SIZE = 5 * 1024 * 1024  # 5MB

def main():
    # Устанавливаем заголовки
    print("Content-Type: application/json; charset=utf-8")
    print("")
    
    if os.environ.get('REQUEST_METHOD') != 'POST':
        print(json.dumps({'error': 'Method not allowed'}))
        sys.exit(0)
    
    # Создаем директорию, если не существует
    try:
        UPLOAD_DIR.mkdir(parents=True, exist_ok=True)
        UPLOAD_DIR.chmod(0o755)
    except Exception as e:
        print(json.dumps({'error': f'Cannot create upload directory: {str(e)}'}))
        sys.exit(0)
    
    try:
        # Парсим multipart/form-data
        form = cgi.FieldStorage()
        
        if 'image' not in form:
            print(json.dumps({'error': 'No image file'}))
            sys.exit(0)
        
        file_item = form['image']
        
        if not hasattr(file_item, 'filename') or not file_item.filename:
            print(json.dumps({'error': 'No file uploaded'}))
            sys.exit(0)
        
        # Читаем файл
        file_data = file_item.file.read()
        
        # Проверяем размер
        if len(file_data) > MAX_FILE_SIZE:
            print(json.dumps({'error': 'File too large'}))
            sys.exit(0)
        
        # Генерируем уникальное имя файла
        age = form.getvalue('age', '0')
        bio_age = form.getvalue('bioAge', '0')
        hash_str = hashlib.md5(f"{age}_{bio_age}_{datetime.now().timestamp()}".encode()).hexdigest()[:12]
        filename = f'share_{hash_str}.png'
        filepath = UPLOAD_DIR / filename
        
        # Сохраняем файл
        with open(filepath, 'wb') as f:
            f.write(file_data)
        
        # Устанавливаем права
        os.chmod(filepath, 0o644)
        
        # Возвращаем публичный URL
        protocol = 'https' if os.environ.get('HTTPS') == 'on' or os.environ.get('HTTP_X_FORWARDED_PROTO') == 'https' else 'http'
        host = os.environ.get('HTTP_HOST', 'app.medicinetest.ru')
        public_url = f"{protocol}://{host}/share-images/{filename}"
        
        print(json.dumps({'url': public_url}))
        
    except Exception as e:
        import traceback
        error_msg = traceback.format_exc()
        print(json.dumps({'error': f'Server error: {str(e)}'}))
        sys.exit(0)

if __name__ == '__main__':
    main()
